packages/server-wallet/src/onchain-service/transaction-submission.ts

```diff
@@ -0,0 +1,154 @@
+  static readonly reasons = {
+    zeroAttempts: 'Invalid max transaction submission attempt count of 0',
+    failedAllAttempts: 'Failed all transaction attempts',
+    unknownError: 'Transaction failed with unkown error',
+  } as const;
+  constructor(
+    reason: Values<typeof TransactionSubmissionError.reasons>,
+    public readonly data: any = undefined
+  ) {
+    super(reason, data);
+    logger.error(reason, data);
+  }
+}
+
+export class TransactionSubmissionService implements TransactionSubmissionServiceInterface {
+  // Chain provider
+  private provider: providers.JsonRpcProvider;
+
+  // Wallet to send transactions
+  private wallet: Wallet;
+
+  // Queues to handle transactions
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Storage for transactions
+  private store: TransactionSubmissionStoreInterface;
+
+  // Memory tracking wallet nonce to assist with retries
+  // TODO: maybe replace with NonceManager?
+  private memoryNonce = 0;
+
+  constructor(
+    provider: string | providers.JsonRpcProvider,
+    wallet: string | Wallet,
+    store: TransactionSubmissionStore
+  ) {
+    this.provider =
+      typeof provider === 'string' ? new providers.JsonRpcProvider(provider) : provider;
+    this.wallet =
+      typeof wallet === 'string'
+        ? new Wallet(wallet, this.provider)
+        : wallet.connect(this.provider);
+
+    this.store = store;
+    logger.log(`Transaction service created`);
+  }
+
+  public async submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction,
+    options: TransactionSubmissionOptions = {}
+  ): Promise<providers.TransactionResponse> {
+    const attempts = options.maxSendAttempts ?? 1;
+    if (attempts === 0) {
+      throw new TransactionSubmissionError(TransactionSubmissionError.reasons.zeroAttempts, {
+        attempts,
+      });
+    }
+
+    const indexedErrors: {[k: string]: string} = {};
+    for (let attempt = 0; attempt < attempts; attempt++) {
+      try {
+        const response = await this.queue.add(() => this._sendTransaction(channelId, tx));
+        logger.log(`Transaction sent`, {hash: response.hash, channelId, to: tx.to});
+        return response;
+      } catch (e) {
+        // Store the error in memory
+        indexedErrors[attempt.toString()] = e.message;
+
+        // Retry IFF known error
+        if (
+          !TransactionSubmissionError.isKnownErr(
+            e.message,
+            Object.values(TransactionSubmissionError.knownErrors)
+          )
+        ) {
+          throw new TransactionSubmissionError(TransactionSubmissionError.reasons.unknownError, {
+            attempt,
+            attempts,
+            indexedErrors,
+          });
+        }
+        logger.warn(`Failed to send tx`, {attempt, error: e.message});
+      }
+    }
+
+    throw new TransactionSubmissionError(TransactionSubmissionError.reasons.failedAllAttempts, {
+      attempts,
+      indexedErrors,
+    });
+  }
+
+  private async _sendTransaction(
```

Comment:
I think either this function, or the `submitTransaction` should probably be tolerant to being called more than once and thus checking storage first before re-submitting

Reply:
So how would you differentiate between the a transaction being re-submitted vs a transaction with the same call data being submitted intentionally while another is in mempool? Is there an assumption we can make that the call data would never be the same? Would we want to make people wait until that transaction has been mined before they are allowed to send a second?

Reply:
Also, this seems more like a v1 issue, where we can do some more creative unique constraints using systems like postgres.

Reply:
(And also store-level transactions)

Reply:
I can't think of any example in the nitro protocol where a transaction would have the exact same calldata of a previous transaction